### PR TITLE
ARROW-3650: [Python] warn on converting DataFrame with mixed type column names

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
https://issues.apache.org/jira/browse/ARROW-3650

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
There is one problem with this PR: it also generates a warning if you have a `None` as column name (so mix of int or string with None in an object dtype Index object), while for this case we actually roundtrip correctly:

```
In [62]: df = pd.DataFrame(np.random.randn(2, 2), columns=['a', None])                                                                                                          

In [63]: pa.Table.from_pandas(df).to_pandas().columns                                                                                                                           
Out[63]: Index(['a', None], dtype='object')
```

But I don't think it is that easy to detect this case, to not warn in that case (I certainly could do it, but IMO would lead to unnecessary code complexity). So it seems it is a choice between no warning at all (as it is currently), or warning in the mixed case including for this specific case where the warning is incorrect. 
I don't feel strongly about either of them (it are corner cases anyway).

**Comment by pitrou** [comment]
Why would someone pass `None` as the column name? What is the expectation?

**Comment by jorisvandenbossche** [comment]
Not sure why someone would ever do that on purpose (also when you have no index name (meaning it is `None`), and you reset the index (meaning: converting it to a column), pandas creates a default index name like 'index' or 'level_0', so you don't end up with a column name `None` in that common case). 

There is a case where you can end up with NaNs in your index or column names (that is not too far stretched), eg when pivoting / unstacking data and your data contains NaNs:

```
In [4]: df = pd.DataFrame({'A' : [0, 0, 1, 1], 'B': ['a', 'b', 'a', np.nan], 'C':[1, 2, 3, 4]})

In [5]: df
Out[5]: 
   A    B  C
0  0    a  1
1  0    b  2
2  1    a  3
3  1  NaN  4

In [6]: df2 = df.pivot(index='A', columns='B', values='C')

In [7]: df2
Out[7]: 
B  NaN    a    b
A               
0  NaN  1.0  2.0
1  4.0  3.0  NaN

In [8]: df2.columns
Out[8]: Index([nan, 'a', 'b'], dtype='object', name='B')

In [9]: table = pa.Table.from_pandas(df2)
/home/joris/miniconda3/envs/arrow-dev/bin/ipython:1: UserWarning: The DataFrame has column
names of mixed type. They will be converted to strings and not roundtrip correctly.


In [10]: table
Out[10]: 
pyarrow.Table
nan: double
a: double
b: double
A: int64
metadata
--------
{b'pandas': b'{"index_columns": ["A"], "column_indexes": [{"name": "B", "field'
            b'_name": "B", "pandas_type": "mixed", "numpy_type": "object", "me'
            b'tadata": null}], "columns": [{"name": "nan", "field_name": "nan"'
            b', "pandas_type": "float64", "numpy_type": "float64", "metadata":'
            b' null}, {"name": "a", "field_name": "a", "pandas_type": "float64'
            b'", "numpy_type": "float64", "metadata": null}, {"name": "b", "fi'
            b'eld_name": "b", "pandas_type": "float64", "numpy_type": "float64'
            b'", "metadata": null}, {"name": "A", "field_name": "A", "pandas_t'
            b'ype": "int64", "numpy_type": "int64", "metadata": null}], "creat'
            b'or": {"library": "pyarrow", "version": "0.13.1.dev130+gdd335952"'
            b'}, "pandas_version": "0.24.2"}'}

In [11]: table.to_pandas()
Out[11]: 
B  nan    a    b
A               
0  NaN  1.0  2.0
1  4.0  3.0  NaN

In [12]: table.to_pandas().columns
Out[12]: Index(['nan', 'a', 'b'], dtype='object', name='B')
```

But apparently (seeing this now as I was trying the above), NaNs indeed don't roundtrip correctly (so in that case the warning is appropriate), and it is only `None` that we reconstruct faithfully:

```
In [17]: df3 = pd.DataFrame([[1, 2], [3, 4]], columns=['a', None])

In [18]: df3.columns
Out[18]: Index(['a', None], dtype='object')

In [19]: table = pa.Table.from_pandas(df3)
/home/joris/miniconda3/envs/arrow-dev/bin/ipython:1: UserWarning: The DataFrame has column
names of mixed type. They will be converted to strings and not roundtrip correctly.

In [20]: table
Out[20]: 
pyarrow.Table
a: int64
None: int64
metadata
--------
{b'pandas': b'{"index_columns": [{"kind": "range", "name": null, "start": 0, "'
            b'stop": 2, "step": 1}], "column_indexes": [{"name": null, "field_'
            b'name": null, "pandas_type": "mixed", "numpy_type": "object", "me'
            b'tadata": null}], "columns": [{"name": "a", "field_name": "a", "p'
            b'andas_type": "int64", "numpy_type": "int64", "metadata": null}, '
            b'{"name": null, "field_name": "None", "pandas_type": "int64", "nu'
            b'mpy_type": "int64", "metadata": null}], "creator": {"library": "'
            b'pyarrow", "version": "0.13.1.dev130+gdd335952"}, "pandas_version'
            b'": "0.24.2"}'}

In [22]: table.to_pandas().columns
Out[22]: Index(['a', None], dtype='object')
```

So here the actual `None` is preserved when converting back to pandas (since we have `"name": null, "field_name": "None"` in the metadata), and the warning is not correct. 
But I *think* having None like that in your column names is less likely than having NaN in your column names.

File: python/pyarrow/pandas_compat.py
Comment (fsaintjacques) [inline python/pyarrow/pandas_compat.py]: ditto

File: python/pyarrow/pandas_compat.py
Comment (fsaintjacques) [inline python/pyarrow/pandas_compat.py]: I'm not familiar, any reason why it's not an equality (`== 'mixed'`) check like the next block? In other words, is the sub-string match desired?

**Review by fsaintjacques — COMMENTED** [review]
I don't think the `None` issue you raised should block this PR.

File: python/pyarrow/pandas_compat.py
Comment (jorisvandenbossche) [reply python/pyarrow/pandas_compat.py]: It was to catch both 'mixed' and 'mixed-integer', IIRC. But can make a comment (or just write it out explicitly in code)

Comment (fsaintjacques) [reply python/pyarrow/pandas_compat.py]: Just wanted to make sure it wasn't a mistake.

Comment (jorisvandenbossche) [reply python/pyarrow/pandas_compat.py]: Added a comment

**Comment by jorisvandenbossche** [comment]
> I don't think the None issue you raised should block this PR.

Yes, I also think that is quite a corner case that is less important. So it should be ready to merge if CI is passing.

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**
